numba/cuda/cudadrv/driver.py

```diff
@@ -268,46 +274,72 @@ def is_available(self):
 
     def __getattr__(self, fname):
         # First request of a driver API function
-        try:
-            proto = API_PROTOTYPES[fname]
-        except KeyError:
-            raise AttributeError(fname)
-        restype = proto[0]
-        argtypes = proto[1:]
-
         self.ensure_initialized()
 
         if self.initialization_error is not None:
             raise CudaSupportError("Error at driver init: \n%s:" %
                                    self.initialization_error)
 
-        # Find function in driver library
-        libfn = self._find_api(fname)
-        libfn.restype = restype
-        libfn.argtypes = argtypes
+        if config.CUDA_USE_CUDA_PYTHON:
+            return self._cuda_python_wrap_fn(fname)
+        else:
+            return self._ctypes_wrap_fn(fname)
+
+    def _ctypes_wrap_fn(self, fname, libfn=None):
+        # Wrap a CUDA driver function by default
+        if libfn is None:
+            try:
+                proto = API_PROTOTYPES[fname]
+            except KeyError:
+                raise AttributeError(fname)
+            restype = proto[0]
+            argtypes = proto[1:]
 
-        safe_call = self._wrap_api_call(fname, libfn)
-        setattr(self, fname, safe_call)
-        return safe_call
+            # Find function in driver library
+            libfn = self._find_api(fname)
+            libfn.restype = restype
+            libfn.argtypes = argtypes
 
-    def _wrap_api_call(self, fname, libfn):
         def verbose_cuda_api_call(*args):
             argstr = ", ".join([str(arg) for arg in args])
             _logger.debug('call driver api: %s(%s)', libfn.__name__, argstr)
             retcode = libfn(*args)
-            self._check_error(fname, retcode)
+            self._check_ctypes_error(fname, retcode)
 
         def safe_cuda_api_call(*args):
             _logger.debug('call driver api: %s', libfn.__name__)
             retcode = libfn(*args)
-            self._check_error(fname, retcode)
+            self._check_ctypes_error(fname, retcode)
+
+        if config.CUDA_LOG_API_ARGS:
+            wrapper = verbose_cuda_api_call
+        else:
+            wrapper = safe_cuda_api_call
+
+        safe_call = functools.wraps(libfn)(wrapper)
+        setattr(self, fname, safe_call)
+        return safe_call
+
+    def _cuda_python_wrap_fn(self, fname):
+        libfn = getattr(cuda_driver, fname)
+
+        def verbose_cuda_api_call(*args):
+            argstr = ", ".join([str(arg) for arg in args])
+            _logger.debug('call driver api: %s(%s)', libfn.__name__, argstr)
+            return self._check_cuda_python_error(fname, libfn(*args))
+
+        def safe_cuda_api_call(*args):
+            _logger.debug('call driver api: %s', libfn.__name__)
+            return self._check_cuda_python_error(fname, libfn(*args))
 
         if config.CUDA_LOG_API_ARGS:
             wrapper = verbose_cuda_api_call
         else:
             wrapper = safe_cuda_api_call
```

Comment:
I felt that this way, although it has a little duplication, is the path of minimal complexity - although it could be factored out some way, the mechanism by which the functions are generated and also how `libfn` is determined differ between the ctypes and Nvidia binding, so it feels like it'd be stuffing multiple disparate things into one function to save a small duplication.

Reply:
Thanks for the explanation. I suppose another advantage of the duplication is that it's trivial to remove the `ctypes` variant at a later time. On the basis of the above, let's leave as-is.